flink-runtime/src/main/java/org/apache/flink/runtime/state/UnexpectedStateHandleException.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.shaded.guava18.com.google.common.base.Joiner;
+
+/**
+ * Signals that an operation did not get the type of {@link StateObject} that was expected. This can
+ * mostly happen when a different {@link StateBackend} from the one that was used for taking a
+ * checkpoint/savepoint is used when restoring.
+ */
+public class UnexpectedStateHandleException extends IllegalStateException {
+
+	@SuppressWarnings("unchecked")
+	public UnexpectedStateHandleException(
+			Class<? extends StateObject> expectedStateHandleClass,
+			Class<? extends StateObject> actualStateHandleClass) {
+		this(new Class[] {expectedStateHandleClass}, actualStateHandleClass);
+	}
+
+	public UnexpectedStateHandleException(
+			Class<? extends StateObject>[] expectedStateHandleClasses,
+			Class<? extends StateObject> actualStateHandleClass) {
```

Comment:
Swap arguments and use ellipse to avoid 2 ctors?

Reply:
I thought about that as well but decided to go with the order that JUnit `assertEquals()` uses. What do you think?